src/hotspot/share/gc/shared/referenceProcessorPhaseTimes.hpp

```diff
@@ -103,14 +103,14 @@ class ReferenceProcessorPhaseTimes : public CHeapObj<mtGC> {
   void print_all_references(uint base_indent = 0, bool print_total = true) const;
 };
 
-class RefProcWorkerTimeTracker : public CHeapObj<mtGC> {
+class RefProcWorkerTimeTracker : public StackObj {
 protected:
   WorkerDataArray<double>* _worker_time;
   double                   _start_time;
   uint                     _worker_id;
 public:
   RefProcWorkerTimeTracker(WorkerDataArray<double>* worker_time, uint worker_id);
-  virtual ~RefProcWorkerTimeTracker();
+  ~RefProcWorkerTimeTracker();
```

Comment:
The usual guideline is that the destructor for a baseclass should be either
public and virtual or protected and non-virtual. This prevents destructor
slicing by `delete`. That the base class in question is a StackObj mitigates
against that kind of error.

And indeed, if I recall correctly, because it's now a StackObj its destructor
can't be virtual, because that would run afoul of the induced "deleting
destructor".

There are some relevant warning options that I occasionally look at and
remember why I think they aren't right for us, so we can't ever enable them
(-Wdelete-non-virtual-dtor and -Wnon-virtual-dtor).

But all that is kind of irrelevant.  The real design problem here is that
RefProcSubPhasesWorkerTimeTracker is derived from RefProcWorkerTimeTracker.
It should instead *have* a RefProceWorkerTimeTracker.  That would eliminate
all the discussion about base class destructors, and would also make
RefProcWorkerTimeTracker no longer need to perform double-duty as both a base
class and as a concrete class (which is often problematic).